### PR TITLE
Add tree tests to ftprime

### PR DESCRIPTION
**Pull request by ashander**
just for now testing the number of trees produced vs supposedly equivalent
msprime simulation

fixing this or getting the scaling right should resolve (some) recombination
issues in https://github.com/petrelharp/ftprime_ms/pull/11

### CODE REVIEW

**Comment by ashander** [comment]
This is failing with my test (see diff) to see if number of trees are within **3** standard errors of the mean number of trees produced in msprime

<details>

<summary> failure examples: showing num trees vs expected (from msprime) </summary>

        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 5.01 < (3 * 0.5627305363721226)
E        +  where 5.01 = abs((13.99 - 19))
E        +    where 19 = <msprime.trees.TreeSequence object at 0x7fbf96e3a240>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 5.800000000000001 < (3 * 0.9375446117331692)
E        +  where 5.800000000000001 = abs((25.8 - 20))
E        +    where 20 = <msprime.trees.TreeSequence object at 0x7fbf83c87b38>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 11.53 < (3 * 0.5409410057641885)
E        +  where 11.53 = abs((14.47 - 26))
E        +    where 26 = <msprime.trees.TreeSequence object at 0x7fbf83cd5a58>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 4.68 < (3 * 0.8740801803130888)
E        +  where 4.68 = abs((26.68 - 22))
E        +    where 22 = <msprime.trees.TreeSequence object at 0x7fbf83cd5a90>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 24.65 < (3 * 1.627564022232196)
E        +  where 24.65 = abs((53.35 - 78))
E        +    where 78 = <msprime.trees.TreeSequence object at 0x7fbf83c563c8>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 6.399999999999999 < (3 * 0.843992436176553)
E        +  where 6.399999999999999 = abs((27.6 - 34))
E        +    where 34 = <msprime.trees.TreeSequence object at 0x7fbf83c45080>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 11.280000000000001 < (3 * 1.505148068154993)
E        +  where 11.280000000000001 = abs((53.72 - 65))
E        +    where 65 = <msprime.trees.TreeSequence object at 0x7fbf83ca3978>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 29.329999999999984 < (3 * 2.412009305307276)
E        +  where 29.329999999999984 = abs((133.67000000000002 - 163))
E        +    where 163 = <msprime.trees.TreeSequence object at 0x7fbf83c596d8>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 12.649999999999977 < (3 * 3.771219426073741)
E        +  where 12.649999999999977 = abs((270.35 - 283))
E        +    where 283 = <msprime.trees.TreeSequence object at 0x7fbf83c57d30>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 2.99 < (3 * 0.5627305363721226)
E        +  where 2.99 = abs((13.99 - 11))
E        +    where 11 = <msprime.trees.TreeSequence object at 0x7fbf83b23e80>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 3.5299999999999994 < (3 * 0.5409410057641885)
E        +  where 3.5299999999999994 = abs((14.47 - 18))
E        +    where 18 = <msprime.trees.TreeSequence object at 0x7fbf83b1d588>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 9.32 < (3 * 0.8740801803130888)
E        +  where 9.32 = abs((26.68 - 36))
E        +    where 36 = <msprime.trees.TreeSequence object at 0x7fbf83b7d518>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 11.44 < (3 * 0.5932822586120693)
E        +  where 11.44 = abs((14.56 - 26))
E        +    where 26 = <msprime.trees.TreeSequence object at 0x7fbf83b23198>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 3.3999999999999986 < (3 * 0.843992436176553)
E        +  where 3.3999999999999986 = abs((27.6 - 31))
E        +    where 31 = <msprime.trees.TreeSequence object at 0x7fbf83c45da0>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 6.280000000000001 < (3 * 1.505148068154993)
E        +  where 6.280000000000001 = abs((53.72 - 60))
E        +    where 60 = <msprime.trees.TreeSequence object at 0x7fbf83adba90>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 18.5 < (3 * 2.936000853218713)
E        +  where 18.5 = abs((132.5 - 151))
E        +    where 151 = <msprime.trees.TreeSequence object at 0x7fbf83adbd30>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 4.5600000000000005 < (3 * 0.5680588957941853)
E        +  where 4.5600000000000005 = abs((14.56 - 10))
E        +    where 10 = <msprime.trees.TreeSequence object at 0x7fbf8303dcc0>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 5.300000000000001 < (3 * 0.8643909515790392)
E        +  where 5.300000000000001 = abs((26.3 - 21))
E        +    where 21 = <msprime.trees.TreeSequence object at 0x7fbf83b2e7f0>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 19.57 < (3 * 1.4345136084852226)
E        +  where 19.57 = abs((54.57 - 35))
E        +    where 35 = <msprime.trees.TreeSequence object at 0x7fbf83bf8c88>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 19.189999999999998 < (3 * 0.8598443598886506)
E        +  where 19.189999999999998 = abs((27.810000000000002 - 47))
E        +    where 47 = <msprime.trees.TreeSequence object at 0x7fbf83563dd8>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 25.630000000000003 < (3 * 1.4632643043013196)
E        +  where 25.630000000000003 = abs((54.63 - 29))
E        +    where 29 = <msprime.trees.TreeSequence object at 0x7fbf83a917b8>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 8.559999999999999 < (3 * 0.9338007487595188)
E        +  where 8.559999999999999 = abs((27.44 - 36))
E        +    where 36 = <msprime.trees.TreeSequence object at 0x7fbf83567358>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 113.47999999999999 < (3 * 2.560953105404472)
E        +  where 113.47999999999999 = abs((134.52 - 248))
E        +    where 248 = <msprime.trees.TreeSequence object at 0x7fbf83726d68>.num_trees
--
        se = std / math.sqrt(len(tree_counts))
    
        assert ts.sample_size == mspts.sample_size
        # TODO fix  expected error
>       assert abs(mean - ts.num_trees) < 3 * se
E       assert 156.26 < (3 * 3.979584770989928)
E        +  where 156.26 = abs((265.74 - 422))
E        +    where 422 = <msprime.trees.TreeSequence object at 0x7fbf83512cf8>.num_trees
</details>

**Comment by ashander** [comment]
@petrelharp can you

- [ ] check my SE of the mean calculation
- [ ] remind me of analytical expectations for number of trees (it'd be great to test against that too!)

**Comment by ashander** [comment]
@petrelharp thanks for earlier feedback . we still have this issue (see two examples above) where our number of trees **exceeds** that from the supposedly equivalent msprime sims. this is the opposite to what we'd expect if simuPOP were skipping recombinations

can you

- [ ] check my choice of tolerance choice/calculation at 3 std devs
- [ ] remind me of analytical expectations for number of trees (it'd be great to test against that too to see where we're going wrong)

File: tests/test_recomb_collector.py [hunk 1]
```diff
@@ -146,6 +148,19 @@ def test_simupop(make_pop, generations, popsize):
...
+                                 recombination_rate=recomb_rate,
+                                 length=max(locus_position))
+        tree_counts.append(mspts.num_trees)
+    mean = 1 / len(tree_counts) * sum(tree_counts)
+    sse = sum((t - mean) ** 2 for t in tree_counts)
+    std = math.sqrt(1 / (len(tree_counts) - 1) * sse)
+
+    assert ts.sample_size == mspts.sample_size
+    # TODO fix  expected error
+    assert abs(mean - ts.num_trees) < 3 * std
```
Comment (petrelharp) [inline tests/test_recomb_collector.py]: this looks like the check you intended.  to verify you could replace ts by another instance of the same `msprime.simulate()` call.

**Comment by petrelharp** [comment]
Since recombs are produced by simuPOP; I'm not sure this should be a ftprime issue.  If there is an issue here, either (a) simuPOP is not producing the number of recombs we expect, or (b) we are not properly recording them. I think we adequately test (b) elsewhere, so a test of this would just involve simuPOP.  It could be done by running simuPOP with N diploid individuals for T generations, writing out recombinations to a file; then the number of recombinations in that file (which something like `cut -f 4- recombs.txt | wc -w`) should be `N * T * r * L` plus/minus `3 * sqrt(N * T * r * L)`, where `r` is the recomb rate per locus and `L` is the number of loci.

**Comment by petrelharp** [comment]
oh, right: ALSO, I am not sure that number of trees is a good check, because I forget if simplify merges adjacent identical edges.

**Comment by ashander** [comment]
Oh! I should check that before I use it any further

**Comment by ashander** [comment]
and good point about this not being inherent to ftprime.recomb_collector. though I'd argue it's nice to have examples where the resulting treeseq matches our input recombination rate! (as you point out, num_trees may not be a good test of this).

> It could be done by running simuPOP with N diploid individuals for T generations, writing out recombinations to a file; then the number of recombinations in that file (which something like cut -f 4- recombs.txt | wc -w) should be N * T * r * L plus/minus 3 * sqrt(N * T * r * L), where r is the recomb rate per locus and L is the number of loci.

but I'll do this

**Pull request merged**